[12.0][FIX] Envío al SII ImporteTotal sin IRPF en facturas de proveedor

JuanjoA-PH opened the pull request:
Al hilo de:  https://github.com/OCA/l10n-spain/issues/1281
He estado trabajando en esto y he modificado el módulo para sacar el irpf del ImporteTotal que se envía al SII, añadiento también un test.
Antes de hacer PR no estaría mal si alguien lo puede probar, por ejemplo con facturas en otra divisa, o se anima a meterle un test para esto.

Añado:
  - Uno de los problemas era como detectar que el impuesto en uso es un IRPF. Esto lo he 'arreglado' buscando en el modelo del impuesto que el campo description contenga la cadena 'irpf'. Aquí no valía buscar por ejemplo  un impuesto negativo. Si veis otra forma más formal, me decís y lo cambio.

JuanjoA committed:
[FIX] l10n_es_aeat_sii: Envío al SII ImporteTotal sin IRPF en facturas de proveedor

- Añadido aeat.sii.map.lines para NotIncludedInTotal
- facturas emitidas irpf

JuanjoA-PH commented:
@pedrobaeza ¿sabes por qué peta travis?

pedrobaeza commented:
```
l10n_es_aeat_sii/models/account_invoice.py:692:49: E713 test for membership should be 'not in'
l10n_es_aeat_sii/models/account_invoice.py:692:89: E501 line too long (95 > 88 characters)
l10n_es_aeat_sii/models/account_invoice.py:697:89: E501 line too long (104 > 88 characters)
```

JuanjoA-PH commented:
Gracias @pedrobaeza ahora corrijo, veía muchos errores en .rst y .js y no me cuadraba.

SoniaViciana commented:
Buenos días @JuanjoA-PH 

Muchas gracias por aportar esta mejora, sin embargo, creo que sería más universal si la condición para no sumar el impuesto en el importe total que informamos al SII, sea mediante un check a nivel de impuestos o plantilla de impuestos, dado que no únicamente encontramos este problema con el IRPF, por ejemplo, existen ciertas operaciones sujetas a **IGIC o IPSI**, que se deben enviar al SII, pero como **exentas**:

![image](https://user-images.githubusercontent.com/46522913/87302782-48d19780-c512-11ea-84c9-0d2816f59b98.png)

Factura con IGIC al 7% que se envía al SII como Exenta conforme a la entrada anterior de la AEAT:

![image](https://user-images.githubusercontent.com/46522913/87303233-1b391e00-c513-11ea-9312-88e80bd6794e.png)

Muchas gracias!

JuanjoA-PH commented:
Hola, hasta dentro de 2 semanas no podré ponerme a revisar esto.

JuanjoA-PH commented:
Hola @SoniaViciana , suena interesante, a ver qué opina @pedrobaeza o alguién por si puede influir en otros módulos. Se lo veis bien, lo cambio.

pedrobaeza commented:
Lo del check no lo veo, porque supone tener que traspasar información de la plantilla del plan contable al plan instalado si quieres que aparezcan los "por defecto".

Justo el caso que comenta Sonia es un caso muy remoto, ya que la mayoría de operaciones con Canarias son sujetas a reglas de localización (que para eso creamos la posición fiscal específica) o con sede permanente.

De todas formas, si se quiere dejar esa posibilidad abierta de incluir otros impuestos, yo como lo implementaría es teniendo un mapeo de impuestos SII extra (`aeat.sii.map.line`) con el IRPF de base (y si aplica a alguno más del plan estándar `l10n_es`), y que excluya los de ese mapeo. Así se tiene un compromiso entre extensión y por defecto sanos. Que se crean nuevos impuestos "a mano", pues habrá que crearlos en la plantilla del plan contable por código en su lugar y ejecutar `account_chart_update`, como otras muchas cosas que ya se hacen así para mantener la coherencia.

JuanjoA-PH commented:
@pedrobaeza entonces según entiendo, crear una nueva entrada en `aeat.sii.map.lines` , añadiendo en data.xml una entrada al menos para el IRPF y a la hora de calcular el ImporteTotal, descontar los impuestos incluidos en esa entrada, ¿es así?

¿Qué nombre sugieres para esa entrada? ¿Directamente IRPF?

pedrobaeza commented:
Sí, eso sería. Yo pondría de clave a eso "NotIncludedInTotal" o similar (no tienen por qué ser código crípticos y además accedemos a ellos de la misma forma tengan un código u otro).

JuanjoA-PH commented:
@SoniaViciana a ver si con estos cambios podéis probar.
Se ha creado una línea del mapeo de impuestos del SII que hace que los impuestos incluidos en esa línea no se incluyan en el total.
En mi caso, el IRPF es negativo y al no incluirlo en el total (ya que está en el mapeo), ya sale correctamente de cara a hacienda.
De hecho, he metido todos los irpf que tenemos por defecto en ese mapeo.

pedrobaeza commented on l10n_es_aeat_sii/README.rst:
```diff
@@ -145,6 +145,7 @@ Contributors
 * Ismael Calvo <ismael.calvo@factorlibre.es>
 * Omar Castiñeira - Comunitea S.L. <omar@comunitea.com>
 * Juanjo Algaz <jalgaz@gmail.com>
+* Juanjo Algaz <juanjoalgaz@planetahuerto.es> - Planeta Huerto
```
No añadas dos veces tu nombre, y además no se edita aquí, si no en el fragmento.

JuanjoA-PH replied:
Jeje, esta sabía que caía, uhmm, algunas aportaciones como persona individual y otras con tiempo de empresa, ¿cómo lo reflejo?, ¿misma línea separado por comas?

pedrobaeza replied:
El contribuidor es el mismo, da igual que lo hagas a título personal o trabajando para una empresa. Lo de poner los mails es algo que yo ya no suelo hacer, porque de hecho recibo un montón de correos de "leechers", pero puedes poner ambos en la misma línea si así lo quieres, o yo directamente solo pondría `Juanjo Algaz` para no complicarme.

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -674,6 +674,39 @@ def _get_account_registration_date(self):
+        Avoid to send IRPF data to SII systems,
+        but only check supplier invoices
+        """
+        taxes_notincludedintotal = self._get_sii_taxes_map(
```
Haz esto solo si es supplier invoice. De todas formas, también existe la retención de cliente. ¿Las mismas hay que incluirlas en el total?

JuanjoA-PH replied:
Consulto en el otro hilo, creo que si, pero como no tengo confirmación no lo cambié.

JuanjoA-PH replied:
Por otro lado, serán muy pocos los casos en los que un emisor de facturas con irpf tenga que declarar al SII, ¿no? en concreto los del REDEME, porque si facturas mas de 6 millones de euros como autónomo o profesional, . . .

pedrobaeza replied:
Bueno, puede haberse acogido voluntariamente al SII, o la idea es que lo hagan obligatorio para todo el mundo... Si procede y lo tenemos contemplado, mejor. Si no, se puede anotar en el ROADMAP por tampoco que esto suponga un bloqueo.

JuanjoA-PH replied:
Listo y probado por @vipvalen (gracias).

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -674,6 +674,39 @@ def _get_account_registration_date(self):
+            lines = [
+                line for line in self.tax_line_ids
+                if line.tax_id not in taxes_notincludedintotal]
+            amount_tax_no_irpf = sum(round_curr(
```
Aquí diría que no habría que redondear inmediatamente. Depende de cómo lo haga el estándar, Si no, hay riesgo de descuadre.  ¿Lo has comprobado?

JuanjoA-PH replied:
Biene del anterior código, por lo que entiendo que está ok, por eso lo dejé.

pedrobaeza replied:
Uhm, ¿dónde dices que estaba así en el anterior código? Antes se cogía el total y se redondeaba, así que no existía ese redondeo. También veo que quitas el `abs`, por lo que podría ser problemático dependiendo de los signos. ¿Lo has comprobado?

JuanjoA-PH replied:
Es del método *_compute_amount* en *addons_shared/OCA/OCB/addons/account/models/account_invoice.py:57* (en mi versión de OCB).
Quizás debería coger el amount que ya hay y restarle los IRPF o impuestos incluidos en el map.line *NotIncludedInTotal*, pero para multidivisa no se si es mejor dejarlo como está, que es recalcularlo sacando esos impuestos.

JuanjoA-PH replied:
Pienso que esto está ok también.

JuanjoA-PH commented:
Se ha implementado la parte cliente, esperando a ver si alguién puede probarlo / validarlo.

pedroortega-punt commented:
@JuanjoA-PH 

Revisión:

Factura con IRPF y con 21% IVA, ok
factura con IRPF y con 21% IVA y con 10% IVA, ok
factura con IRPF arrendamientos y con 21% IVA, ok
Rectificativa con IRPF y con 21% IVA, ok

JuanjoA-PH commented:
Gracias @vipvalen , a ver si hoy puedo revisar el coverage y se aprueba el PR.

JuanjoA-PH commented:
@pedrobaeza creo que esto ya está.

HaraldPanten commented:
¿Te encajan los cambios @pedrobaeza?  Lo digo para tenerlo en cuenta en la migración a la V13

Saludos.

SoniaViciana commented:
Buenos días @JuanjoA-PH @pedrobaeza 

Hemos realizado un backport a v11 para realizar pruebas y nos ha funcionado perfectamente, en ningún caso se suma al importe total informado en el SII aquellas cuotas de impuestos incluidos en la fila de mapeo "NotIncludedInTotal".

Muchas gracias!

JuanjoA-PH commented:
@pedrobaeza ¿ves ok el PR para integrarlo?

JuanjoA-PH commented:
Hola @pedrobaeza ¿se puede integrar esto?
Gracias

pedrobaeza commented:
Estoy revisándolo ahora de cara a pasar todo a v13. ¿Estás por ahí para respuesta rápida si es necesario :smiley: ?

pedrobaeza commented on l10n_es_aeat_sii/README.rst:
```diff
@@ -23,7 +23,7 @@ Suministro Inmediato de Información en el IVA
     :alt: Try me on Runbot
 
-|badge1| |badge2| |badge3| |badge4| |badge5| 
+|badge1| |badge2| |badge3| |badge4| |badge5|
```
Estos cambios no deberían estar y es por editar a mano con un IDE que hace trim de los espacios finales.

JuanjoA-PH replied:
Corrijo en un ratín :-) Si, estoy, liadillo pero si es para algo concreto, lo miro, gracias!

pedrobaeza reviewed (changes requested):
Aunque formalmente está correcto, la implementación no la veo la más adecuada porque hay que procesar dos veces los mismos conceptos. Sería mejor cambiar `_get_sii_out_taxes` de forma totalmente análoga a `_get_sii_in_taxes` para que acumule y devuelva dos datos en lugar de uno con el importe total y ese diccionario. En `_get_sii_in_taxes` habrá que devolver 3 en lugar de 2 también.

pedrobaeza commented:
Dime si puedes hacerlo o lo hago yo.

JuanjoA-PH commented:
@pedrobaeza pues si le puedes meter mano tú, te lo agradezco, mañana queremos sacar un tema y estamos a tope.

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_sii: Adjust "total without IRPF" algorithm

- Don't process again everything, but do it inline on existing methods
- Make tests resilient

pedrobaeza commented:
@JuanjoA-PH echa un vistazo para que veas cómo queda

JuanjoA-PH commented:
Ok @pedrobaeza los veo correctos y entiendo el cambio. El tema es si salen nuevas casuísticas que obliguen a devolver más valores desde los métodos, pero así nos ahorramos iteraciones.
Gracias por los cambios.

pedrobaeza commented:
Sí, conforme nos surjan, vamos valorando opciones. Si es necesario devolver más cosas, ya pasaríamos a devolver un diccionario y no tenemos que desdoblar variables.

/ocabot merge patch
